src/main/java/net/dries007/tfc/client/model/PlantBlockModel.java

```diff
@@ -0,0 +1,309 @@
+import net.neoforged.neoforge.client.model.data.ModelProperty;
+import net.neoforged.neoforge.client.model.geometry.IGeometryBakingContext;
+import net.neoforged.neoforge.client.model.geometry.IGeometryLoader;
+import org.jetbrains.annotations.Nullable;
+import net.minecraft.client.renderer.block.model.BlockModel;
+import net.neoforged.neoforge.client.model.IDynamicBakedModel;
+import net.neoforged.neoforge.client.model.geometry.IUnbakedGeometry;
+
+import net.dries007.tfc.TerraFirmaCraft;
+import net.dries007.tfc.client.ClientHelpers;
+import net.dries007.tfc.client.ClimateRenderCache;
+import net.dries007.tfc.client.RenderHelpers;
+import net.dries007.tfc.common.blocks.plant.Plant;
+import net.dries007.tfc.common.blocks.plant.PlantBlock;
+import net.dries007.tfc.util.EnvironmentHelpers;
+import net.dries007.tfc.util.Helpers;
+import net.dries007.tfc.util.calendar.Calendars;
+import net.dries007.tfc.util.climate.Climate;
+import net.dries007.tfc.util.climate.OverworldClimateModel;
+import net.dries007.tfc.util.registry.RegistryPlant;
+import net.dries007.tfc.world.settings.Settings;
+
+public class PlantBlockModel implements IDynamicBakedModel, IUnbakedGeometry<PlantBlockModel>
+{
+    private final BlockModel dormant;
+    private final BlockModel sprouting;
+    private final BlockModel budding;
+    private final BlockModel blooming;
+    private final BlockModel seeding;
+    private final BlockModel dying;
+
+    @Nullable private BakedModel dormantBakedModel;
+    @Nullable private BakedModel sproutingBakedModel;
+    @Nullable private BakedModel buddingBakedModel;
+    @Nullable private BakedModel bloomingBakedModel;
+    @Nullable private BakedModel seedingBakedModel;
+    @Nullable private BakedModel dyingBakedModel;
+
+    public PlantBlockModel(BlockModel dormant, BlockModel sprouting, BlockModel budding, BlockModel blooming, BlockModel seeding, BlockModel dying)
+    {
+        this.dormant = dormant;
+        this.sprouting = sprouting;
+        this.budding = budding;
+        this.blooming = blooming;
+        this.seeding = seeding;
+        this.dying = dying;
+    }
+
+    @Override
+    public ModelData getModelData(BlockAndTintGetter level, BlockPos pos, BlockState state, ModelData data)
+    {
+        return data.derive().with(BakedModelData.PROPERTY, new BakedModelData(getModelFromBlockState(state, pos))).build();
+    }
+
+    private BakedModel getModelFromBlockState(@Nullable BlockState state, @Nullable BlockPos pos)
+    {
+        if (pos == null)
+        {
+            pos = BlockPos.ZERO;
+        }
+        if (state == null)
+        {
+            return getModelFromCalendar();
+        }
+        final Block block = state.getBlock();
+        if (!(block instanceof PlantBlock))
+        {
+            return getModelFromCalendar();
+        }
+        else
+        {
+            final RegistryPlant plant = ((PlantBlock) block).getPlant();
+            float start = plant.getBloomOffset();
```

Comment:
Little confused what you mean here, as in I should re-organize this so that `start` is multiple `final` variables rather than one that gets modified? Or are you saying that `plant` shouldn't be final?